### PR TITLE
[Twig] Add param & return type hints to all files in the Twig dir

### PR DESCRIPTION
**Pull request by tambait**
## Subject

This is a follow up of #6404 to add param & return type hints.

## Changelog

```markdown
### Added
- Added parameter and return type hints to all files in the Twig dir

### CODE REVIEW

**Comment by tambait** [comment]
When I run phpstan analysis (I'm using level 6) I get messages about missing iterable value type, for example

```
Method Sonata\AdminBundle\Twig\Extension\SonataAdminExtension::renderListElement() has parameter $listElement with no value type specified in  iterable type array.                                                                                                                                 
💡 Consider adding something like array<Foo> to the PHPDoc. 
```
Should I ignore these messages for now and focus just on adding type hints?

**Comment by VincentLanglet** [comment]
> When I run phpstan analysis (I'm using level 6) I get messages about missing iterable value type, for example
> 
> ```
> Method Sonata\AdminBundle\Twig\Extension\SonataAdminExtension::renderListElement() has parameter $listElement with no value type specified in  iterable type array.                                                                                                                                 
> 💡 Consider adding something like array<Foo> to the PHPDoc. 
> ```
> 
> Should I add these types to the PHPDoc block or I can ignore messages for now and focus just on adding type hints?

You don't have to, but if you do it, it's welcomed.

**Review by VincentLanglet — DISMISSED** [review]

File: src/Twig/Extension/SonataAdminExtension.php
Comment (VincentLanglet) [inline src/Twig/Extension/SonataAdminExtension.php:164]:
```suggestion
        array $params = []
```

File: src/Twig/Extension/SonataAdminExtension.php
Comment (VincentLanglet) [inline src/Twig/Extension/SonataAdminExtension.php:205]: You should remove this to avoid phpstan error

Comment (tambait) [reply src/Twig/Extension/SonataAdminExtension.php]: Thanks. There are different parameters and return types which are mixed etc, but I left those as they are. Not sure what's good practice in those cases, just to document them in phpdoc block? Maybe I'll g add missing iterable value types and method descriptions at some point in future if I get more familiar with the code and now to with adding just type hints so to move forward with the tasks needed for S5.

Comment (VincentLanglet) [reply src/Twig/Extension/SonataAdminExtension.php]: I'm not sure to understand.

When mixed it the valid value (because we accept everything) we should use mixed in phpdoc.
When it's not, (like in your case, you added `object` typehint)  we should not use mixed.

Some `mixed` are wrong, and we try to fix them with these PR about param/return typehint.

File: src/Twig/Extension/SonataAdminExtension.php
Comment (VincentLanglet) [inline src/Twig/Extension/SonataAdminExtension.php:147]: This is redundant with the return type hint.

Why phpcs fixer is not complaining...

File: src/Twig/Extension/SonataAdminExtension.php
Comment (VincentLanglet) [inline src/Twig/Extension/SonataAdminExtension.php:376]: The phpdoc is not needed anymore

**Review by VincentLanglet — APPROVED** [review]

File: src/Twig/Extension/SonataAdminExtension.php
Comment (greg0ire) [inline src/Twig/Extension/SonataAdminExtension.php:87]: Please contribute this to 3.x, and anything else that is backward-compatible

**Review by greg0ire — DISMISSED** [review]

File: src/Twig/Extension/SonataAdminExtension.php
Comment (VincentLanglet) [reply src/Twig/Extension/SonataAdminExtension.php:87]: Cant we ask for a PR on 3.x without blocking this one ?

The last time we did this, it was with a PR of @jordisala1991 and @core23.
It delayed a lot the Pr on master because of the conflict.

When our goal is to release Sonata 4, Im dont see as needed to report everything to 3.x (especially when new contributors are helping us) but I'm maybe the only one to think this way.

Comment (VincentLanglet) [reply src/Twig/Extension/SonataAdminExtension.php:87]: TLDR: to me a Pr on 3.x is a bonus point, not a blocking point

Comment (greg0ire) [reply src/Twig/Extension/SonataAdminExtension.php:87]: To me it's blocking: didn't you complain there were to many merge up conflicts the other day? Contributing things to `master` that could be contributed to `3.x` is a recipe for conflicts. To me the workflow should be
- contribute BC things to `3.x`
- wait for the merge up (the contributor can ask us to speed that up if they are eager to contribute)
- contribute non-BC things to `master`.

Comment (jordisala1991) [reply src/Twig/Extension/SonataAdminExtension.php:87]: > To me it's blocking: didn't you complain there were to many merge up conflicts the other day? Contributing things to `master` that could be contributed to `3.x` is a recipe for conflicts. To me the workflow should be
> 
> - contribute BC things to `3.x`
> 
> - wait for the merge up (the contributor can ask us to speed that up if they are eager to contribute)
> 
> - contribute non-BC things to `master`.

That workflow delay a lot the work on sonata 4 IMO

Comment (VincentLanglet) [reply src/Twig/Extension/SonataAdminExtension.php:87]: I did complain about conflict from merge up. But Im not sure this workflow will help.

The more we're doing PR on 3.x the more we will have conflict because of deprecated method, phpdoc already updated, ...

When core23 followed your workflow, I ended up with lot of conflict during the merge and jordi had lot of conflict during the rebase. 

And I agree with jordi that it delays Sonata 4. Which urge, because we're still not Symfony 5 compatible.

I would prefer to not delay contributor contribution on Master because we're not a lot working on this...

Comment (greg0ire) [reply src/Twig/Extension/SonataAdminExtension.php:87]: > The more we're doing PR on 3.x the more we will have conflict because of deprecated method, phpdoc already updated, ...

Conflicts happen because branches diverge, this makes no sense to me. While I'd agree that you cannot have conflicts if you have never anything to merge up, if anyone decides to contribute some of the changes in this PR, but slightly different on 3.x, you will have a conflict. So in the long run, not doing most of the work on 3.x results in more conflicts. If you are not convinced, please try merging up doctrine/orm 2.8.x into master, you will see what I mean.

> When core23 followed your workflow, I ended up with lot of conflict during the merge and jordi had lot of conflict during the rebase.

Because that workflow was not followed perfectly before, maybe? That might also happen because 3.0.0 is very old.

> And I agree with jordi that it delays Sonata 4. Which urge, because we're still not Symfony 5 compatible.

Anyway, while the choices you are making are harmful in the long run IMO, your argument that there won't be a long run because 4.0.0 is happening soon is valid, and you are more involved than I in fulfilling that goal. I will trust your judgement and dismiss my review.

Once 4.0.0 is out though, let's try again following a workflow that strongly discourages contributing anything backwards compatible to master, otherwise we will end in the same painful situation.

Comment (VincentLanglet) [reply src/Twig/Extension/SonataAdminExtension.php:87]: > Once 4.0.0 is out though, let's try again following a workflow that strongly discourages contributing anything backwards compatible to master, otherwise we will end in the same painful situation.

Im 100% with you on this.
Imho the first thing to do will be to release major more often.

**Review by VincentLanglet — APPROVED** [review]

**Review by VincentLanglet — DISMISSED** [review]

**Review by wbloszyk — APPROVED** [review]

**Comment by VincentLanglet** [comment]
Same here, rebase master then I'll merge this PR @tambait :)

**Review by VincentLanglet — APPROVED** [review]

**Comment by VincentLanglet** [comment]
Thanks @tambait

**Pull request merged**
